spring-kafka/src/test/java/org/springframework/kafka/listener/KafkaMessageListenerContainerTests.java

```diff
@@ -560,6 +565,161 @@ public void testBatchAck() throws Exception {
+		containerProps.setMessageListener((BatchMessageListener<Integer, String>) messages -> {
+			logger.info("batch listener: " + messages);
+		});
+		containerProps.setSyncCommits(true);
+		containerProps.setAckMode(AckMode.BATCH);
+		containerProps.setPollTimeout(10000);
+		containerProps.setAckOnError(false);
+
+		KafkaMessageListenerContainer<Integer, String> container = new KafkaMessageListenerContainer<>(cf,
+				containerProps);
+		container.setBeanName("testBatchListener");
+		container.start();
+		Consumer<?, ?> containerConsumer = spyOnConsumer(container);
+		final CountDownLatch firstBatchLatch = new CountDownLatch(1);
+		final CountDownLatch latch = new CountDownLatch(2);
+		willAnswer(invocation -> {
+
+			@SuppressWarnings({ "unchecked" })
+			Map<TopicPartition, OffsetAndMetadata> map = (Map<TopicPartition, OffsetAndMetadata>) invocation
+					.getArguments()[0];
+			for (Entry<TopicPartition, OffsetAndMetadata> entry : map.entrySet()) {
+				if (entry.getValue().offset() == 2) {
+					firstBatchLatch.countDown();
+					latch.countDown();
+				}
+			}
+			return invocation.callRealMethod();
+
+		}).given(containerConsumer)
+				.commitSync(any());
+
+		assertThat(firstBatchLatch.await(9, TimeUnit.SECONDS)).isTrue();
+
+		assertThat(latch.await(60, TimeUnit.SECONDS)).isTrue();
+		Consumer<Integer, String> consumer = cf.createConsumer();
+		consumer.assign(Arrays.asList(new TopicPartition(topic8, 0), new TopicPartition(topic8, 1)));
+		assertThat(consumer.position(new TopicPartition(topic8, 0))).isEqualTo(2);
+		assertThat(consumer.position(new TopicPartition(topic8, 1))).isEqualTo(2);
+		container.stop();
+		consumer.close();
+		logger.info("Stop batch listener");
+	}
+
+	@Test
+	public void testBatchListenerManual() throws Exception {
+		logger.info("Start batch listener manual");
+
+		Map<String, Object> senderProps = KafkaTestUtils.producerProps(embeddedKafka);
+		ProducerFactory<Integer, String> pf = new DefaultKafkaProducerFactory<>(senderProps);
+		KafkaTemplate<Integer, String> template = new KafkaTemplate<>(pf);
+		template.setDefaultTopic(topic9);
+		template.sendDefault(0, 0, "foo");
+		template.sendDefault(0, 0, "baz");
+		template.sendDefault(1, 0, "bar");
+		template.sendDefault(1, 0, "qux");
+		template.flush();
+
+		Map<String, Object> props = KafkaTestUtils.consumerProps("test9", "false", embeddedKafka);
+		props.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
+		DefaultKafkaConsumerFactory<Integer, String> cf = new DefaultKafkaConsumerFactory<>(props);
+		ContainerProperties containerProps = new ContainerProperties(topic9);
+		final CountDownLatch latch = new CountDownLatch(4);
+		containerProps.setMessageListener((BatchAcknowledgingMessageListener<Integer, String>) (messages, ack) -> {
+			logger.info("batch listener manual: " + messages);
+			for (int i = 0; i < messages.size(); i++) {
+				latch.countDown();
+			}
+			ack.acknowledge();
```

Comment:
This is definitely a race condition.
You `countDown()` to zero and let barrier to overcome, but you send commits only after that.
So, all those assertions in the end are going to fail from time to time.